KEP #77 Dynamically Sized Jobs (#1851)
kubernetes-sigs/kueue: keps/77-dynamically-sized-jobs/README.md

- Resizing of RayJobs
- Partial Preemption

## Proposal

Update the Job framework reconciler and introduce new controllers to orchestrate dynamic resizing of jobs. We are only interested in horizontal scaling of jobs (e.g. scaling more replicas). At a high level, this will be accomplished by:
- Creating Workload Slice objects that represent incremental scale up of jobs. This gives us per-replica control of workload admission. Workload Slices will be garbage collected and consolidated with their parent workloads after successful admission.
- Adding default scheduling gates to control the scheduling of new pods based on their admission status.
- Dynamically adjust quotas in ClusterQueue based on scaling events.

For the MVP, we will only focus on admission of **RayClusters** with autoscaling enabled.

### User Stories

#### Story 1 - RayCluster w/ autoscaling

1. The user creates a RayCluster with `enableInTreeAutoscaling: true`.
2. Kueue admits the RayCluster based on the requested resources in the head pod and worker pod. 
3. Given the deman of the job, the Ray Autoscaler adjusts the replicas field as it adds or removes Pods from the cluster.

> 4. Kueue will not suspend and requeue the RayCluster, instead it will dynamically update ClusterQueue usage based on the scaling event. 

Reviewer: What if the WorkloadSlice is not admitted for a prolonged period of time? Can RayCluster still progress or can we run into a deadlock situation?

Reviewer 2: > What if the WorkloadSlice is not admitted for a prolonged period of time? Can RayCluster still progress or can we run into a deadlock situation?

From our testing so far, there are no deadlock situations as long as KubeRay can see the Pods per replica. This is why we took the approach of using scheduling gates so that KubeRay can observe pods belonging to a RayCluster.

### Notes/Constraints/Caveats (Optional)

If Kueue needs to preempt the resized RayCluster, it would preempt it as a whole, regardless of whether the RayCluster was previously scaled up.

> ## Design Details

Reviewer 3: Observability-wise, generalising the design to other kind of workload that support elasticity, would it be useful for example to have a condition on the parent workload that captures the status of the scaling progression, e.g. in a similar way the Deployment controller does?

Reviewer 4: That is a good point. I guess the workload reconciler could do this based on whether it sees any dependent Workloads.

Reviewer 4: @vicentefb can you add the "ResizeRequested" condition in the parent Workload to the design?

Author: Sg, added.

### Workload Slices

> To support horizontal scaling of jobs, we will introduce the concept of a "Workload Slice”. A Workload Slice is a Workload object with an owner reference to the original Workload for a job. Workload Slices represent per-replica changes to a job that were not initially accounted for when the job was created.

Reviewer 2: > Do the workload slices need to be submitted to the same LocalQueue, or just the same ClusterQueue is enough?

Because LocalQueue is referenced by the top-level RayCluster, it will be submitted to the same LocalQueue

> Do the workload slices need to use identical PodTemplates?

Yes, this is already enforced by KubeRay

> If the two need to be true, what if there is a workload slice that violates the constraints?

From our initial prototyping so far, I think it will be hard to break these constraints. Both the LocalQueue and Pod template are referenced top-level in the RayCluster API.

> 3. Are the workload slices admitted, by Kueue scheduler. to the same resource flavors? (I suppose this is required so that we can aggregate them)

Yes, I think from discussions with Aldo we agreed that all workload slices need to be belong to the same resource flavor

> To support horizontal scaling of jobs, we will introduce the concept of a "Workload Slice”. A Workload Slice is a Workload object with an owner reference to the original Workload for a job. Workload Slices represent per-replica changes to a job that were not initially accounted for when the job was created.

Reviewer 5: > Yes, this is already enforced by KubeRay

Are the Workload slices going to be created by KubeRay?

> From our initial prototyping so far, I think it will be hard to break these constraints. Both the LocalQueue and Pod template are referenced top-level in the RayCluster API.

The constraints probably can be violated if a user creates the Workload slices by hand. Might be worth a note how we handle a mismatch.

> To support horizontal scaling of jobs, we will introduce the concept of a "Workload Slice”. A Workload Slice is a Workload object with an owner reference to the original Workload for a job. Workload Slices represent per-replica changes to a job that were not initially accounted for when the job was created.

Reviewer 4: prebuilt workload feature for context #1575

The benefit of Workload Slices is that Kueue can evaluate admission on a per-replica basis without changing the existing semantics of the Workload API. Once a Workload Slice is admitted, it will be garbage collected and its resources will be aggregated into the admission status of the parent workload.

> ### Creating Workload Slices

Reviewer: What if a RayCluster changes its size again before the previously created WorkloadSlice is admitted? Will we create an additional delta WorkloadSlice (say +5 and +3) or the existing WorkloadSlice will be modified?

Reviewer 2: > Will we create an additional delta WorkloadSlice (say +5 and +3) or the existing WorkloadSlice will be modified?

Additional deltas will be created, but there will be some logic added to deleted excess WorkloadSlice objects (either in existing controller or a new one)

Reviewer: Why not updating the already existing workload slice for scale ups ? For scale downs updates will have to happen anyway.

Reviewer 3: My understanding is that scale ups beyond the currently admitted workload, have to go through queueing and quota reservation, and updating the existing workload slice might play against the sorting policy?

Reviewer: I'm talking about non-admitted workload slice, not the "main" already-admitted workload with which the slice is merged once accepted. It the slice is not yet admitted I don't see a big harm in allowing it to increase if we allow to decrease it.

Reviewer 3: Ah right. It would be consistent with how updates to pending workloads are handled today. AFAIK workloads are updated in-place in that case when compute requests for their jobs are increased. As long as we are sure it'd be correct w.r.t. sorting / queueing policies, it looks like it'd be a good simplification.

Reviewer 4: There is a benefit to having multiple slices: Kueue can admit one of the slices at a time, which could be handy in an already busy cluster.

The [GenericJob interface](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobframework/interface.go#L30-L55) will be updated to handle resize operations of jobs.

```golang
type GenericJob interface {
  ...
  ...
  ResizeJob(wl *kueue.Workload) error
}
```

> Jobs implementing the ResizeJob method will create a Workload Slice for every new replica of a job. 

Reviewer 3: My understanding is a workload slice is only created in case the number of replicas exceeds that of the latest consolidated / admitted workload state. That is if the workload successfully scaled to N replicas, then scaled down to N-m, and then scaled back to N, no slices will be created in that latter case. Is my understanding correct?

Reviewer 2: > That is if the workload successfully scaled to N replicas, then scaled down to N-m, and then scaled back to N, no slices will be created in that latter case. Is my understanding correct?

Generally yes, but in practice it will depend on the time between scale down to M and scale up to N. If there's a long enough gap, on scale down to M we will change the original Workload and then on scale up to N we will create N-M WorkloadSlice objects that go through admission and scheduling gates.

Reviewer 3: > If there's a long enough gap, on scale down to M we will change the original Workload's resources and then on scale up to N we will create N-M WorkloadSlice objects that go through admission and scheduling gate checks

Oh so you're saying that a workload can "loose" its original (as per on creation / first admission) quota on scale down. I think that's worth mentioning it explicitly.

Author: Added

### Pod Scheduling Gates

Inside **raycluster_webhook** implement schedulingGate injection for pods on RayCluster creation time.
The Pods will be ungated following a similar behavior as to how a job is suspended and then unsuspended in the when admitted.
When the RayCluster scales up, the new pods will be gated due to the schedulingGates injection in the webhook.

> After the creation of each individual Workload Slice and admission of a Workload Slice, the **workload_scheduling_gates_controller** should be in charge of removing the scheduling gates from each pod. We only need to ungate the number of pods to match the number of admitted pods, this should be a counter. We don’t want to accidentally ungate too many pods since race conditions could happen and we also don’t want to double count. It's worth mentioning that for the case of recreated pods (i.e. machine failure for example), these pods will go through the admission/scheduuling check again, Kueue is responsible fo removing the scheduling gates when there's available quota and resources to spend on the RayCluster. 

Reviewer: Are all pods created by RayCluster identical?  If not, how will we tell recreated pods that should be admitted immediately from "additional" pods that should wait?

Reviewer 2: > Are all pods created by RayCluster identical?

All worker pods from the same worker group share the same pod template, so it doesn't really matter if newly created pods or older pods are admitted.

Reviewer: Cool :). Please add this information to the KEP (if it is not there already).

Author: Added

Given these changes, check if the delta is positive or negative indicating a scaleup/scaledown. If it’s a scaledown the reconciler should trigger an update on the workload to match the new job’s spec and no quota needs to be checked or accounted for, the cluster queue should update the workload resource usage. 

### Phase 2 - Aggregating Workload Slices

In Phase 2, aggregating Workload Slices into the parent workload will be implemented. This doesn't represent a usable feature to end users, but it can be reviewed independently from the phase 3.

### Phase 3 - Scale up with Workload Slices and Scheduling Gates

In Phase 3, scale up will be implemented by introducing Workload Slices and adding Pod scheduling gates as part of Kueue’s mutating admission for RayCluster. 

When the RayCluster scales, the RayCluster webhook would be modified to intercept and "gate" all pods. Every time there’s a resize, you create a dependable (child) workload slice and once it's admitted, you increase the count in the original workload, delete the old workload and remove the schedulingGates. 
- Pros: You are able to hold the pods added by Kuberay

> - Cons: The fact of having schedulingGates, means we need an API call for every pod, because all pods that are created by the RayCluster are going to have schedulingGates. We need to remove those gates and for every pod you need to make API calls.

Reviewer 3: I understand the gates will be added by the RayCluster webhook to the RayCluster worker templates, and removed individually from the Pods by the workload slices controller. That requires the RayCluster controller in KubeRay to play nice and not overriding Kueue changes, right?

Reviewer 2: >  That requires the RayCluster controller in KubeRay to play nice and not overriding Kueue changes, right? 

That's right, and as far as I'm aware, KubeRay does not mutate scheduling gates in any way.

Reviewer 4: FYI @kevin85421 :)

#### Scheduler

> Since every scale up will have its own individual workload they should proceed to the current scheduling cycle and continue the normal admission process.

Reviewer 3: Is there any impact / need on prometheus metrics? Should the individual workload slice be counted as any other kind of workload, or would it be useful to differentiate them?

Author: For metrics it would be useful to have a breakdown for which Workloads are "real" and which ones are for scale up. We can add a label to the Workload Slice to identify this, maybe something like `PendingWorkload` or `ResizeWorkload`.

  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
- [Design Details](#design-details)
  - [Workload Slices](#workload-slices)
  - [Creating Workload Slices](#creating-workload-slices)
  - [Pod Scheduling Gates](#pod-scheduling-gates)
  - [Garbage Collecting Workload Slices](#garbage-collecting-workload-slices)
- [Phases for MVP (alpha)](#phases-for-mvp-alpha)
  - [Phase 1 - Scale Down](#phase-1---scale-down)
    - [Job controller](#job-controller)
  - [Phase 2 - Aggregating Workload Slices](#phase-2---aggregating-workload-slices)
  - [Phase 3 - Scale up with Workload Slices and Scheduling Gates](#phase-3---scale-up-with-workload-slices-and-scheduling-gates)
    - [Scheduler](#scheduler)
- [Additional Details](#additional-details)
  - [Test Plan](#test-plan)
    - [Unit Tests](#unit-tests)
    - [Integration tests](#integration-tests)
  - [Graduation Criteria](#graduation-criteria)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
  - [Ignore Resize from Kuberay](#ignore-resize-from-kuberay)
<!-- /toc -->

## Summary

> Enable dynamic sizing of Kueue jobs. For the MVP, we will only focus on supporting RayClusters with autoscaling enabled, but other resources that can benefit from dynamic sizing should be supported eventually.

Reviewer 2: yup +1 -- we definitely want this working with all job types that can support resize before going beta / GA

> Enable dynamic sizing of Kueue jobs. For the MVP, we will only focus on supporting RayClusters with autoscaling enabled, but other resources that can benefit from dynamic sizing should be supported eventually.

Reviewer 6: I'd like to rephrase the description in the "Summary" and "Goals" sections. Currently, it looks like focusing on RayJob in this KEP.

- Creating Workload Slice objects that represent incremental scale up of jobs. This gives us per-replica control of workload admission. Workload Slices will be garbage collected and consolidated with their parent workloads after successful admission.
- Adding default scheduling gates to control the scheduling of new pods based on their admission status.
- Dynamically adjust quotas in ClusterQueue based on scaling events.

For the MVP, we will only focus on admission of **RayClusters** with autoscaling enabled.

### User Stories

#### Story 1 - RayCluster w/ autoscaling

1. The user creates a RayCluster with `enableInTreeAutoscaling: true`.
2. Kueue admits the RayCluster based on the requested resources in the head pod and worker pod. 
3. Given the demand of the job, the Ray Autoscaler adjusts the replicas field as it adds or removes Pods from the cluster.
4. Kueue will not suspend and requeue the RayCluster, instead it will dynamically update ClusterQueue usage based on the scaling event. 

### Notes/Constraints/Caveats (Optional)

> If Kueue needs to preempt the resized RayCluster, it would preempt it as a whole, regardless of whether the RayCluster was previously scaled up.

Author: I think an example about when the kueue manager needs to preempt a RayCluster is in partial preemption: What if Kueue needs to free some space to run a higher priority job or to enforce fair sharing? In the current model, the whole RayCluster would be preempted.

> If Kueue needs to preempt the resized RayCluster, it would preempt it as a whole, regardless of whether the RayCluster was previously scaled up.

Reviewer 6: That makes sense. But I'm still wondering how to adapt DyanmicJob to MultiKueue.

## Design Details

> ### Workload Slices

Author: Correct, the Workload Slices aren't CRDs, so we could treat them as internal states, will add that detail.

To support horizontal scaling of jobs, we will introduce the concept of a "Workload Slice”. A Workload Slice is a Workload object with an owner reference to the original Workload for a job. Workload Slices represent per-replica changes to a job that were not initially accounted for when the job was created.

> The benefit of Workload Slices is that Kueue can evaluate admission on a per-replica basis without changing the existing semantics of the Workload API. Once a Workload Slice is admitted, it will be garbage collected and its resources will be aggregated into the admission status of the parent workload.

Reviewer 6: For MultiKueue, the Workload Slices will be created only in the management cluster, right?

Reviewer: In MK slices need to go to 2 clusters - management and the worker that is running the main part of the workload.

Reviewer 4: @vicentefb please add this note about MultiKueue. Just like the slices would go into the same flavor, they would also go into the same cluster in a multi-cluster environment.

Author: Added.

> The benefit of Workload Slices is that Kueue can evaluate admission on a per-replica basis without changing the existing semantics of the Workload API. Once a Workload Slice is admitted, it will be garbage collected and its resources will be aggregated into the admission status of the parent workload.

Reviewer 4: I think this is already being discussed below? https://github.com/kubernetes-sigs/kueue/pull/1851/files#r1537478474

> The benefit of Workload Slices is that Kueue can evaluate admission on a per-replica basis without changing the existing semantics of the Workload API. Once a Workload Slice is admitted, it will be garbage collected and its resources will be aggregated into the admission status of the parent workload.

Reviewer 6: Oh, you're correct. Thank you for pointing that out!

- Workload Slices will be submitted to the same LocalQueue that's referenced by the top-level RayCluster
- Workload Slices will be created by Kueue and use identical PodTemplates which is already enforced by Kuberay
- Workload Slices will beneed to belong to the same resource flavor as the top-level RayCluster that was initially admitted

> ### Creating Workload Slices

Reviewer 6: Could you clarify how to keep compatibility when the partial admission occurs during the first admission?

Reviewer 4: There is no partial admission for RayClusters.

However, in general, I don't see a problem. The job can still request a resize to match their original requested total size or bigger.

There are some implementation considerations, such as paying attention to the `.count` fields in the Workload status. But I don't see anything particularly problematic.

Reviewer 6: > There is no partial admission for RayClusters.

Yes, that's right. But I'd like to clarify it for batch/v1 Job integration in the future.

> However, in general, I don't see a problem. The job can still request a resize to match their original requested total size or bigger.
> 
> There are some implementation considerations, such as paying attention to the .count fields in the Workload status. But I don't see anything particularly problematic.

That makes sense generally, but what happens when requests are submitted with scale-in counts less than minCount? I'd like to clarify the minCount semantics in this situation.

> The [GenericJob interface](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobframework/interface.go#L30-L55) will be updated to handle resize operations of jobs.

Author: Done

```golang
type GenericJob interface {
  ...
  ...
  ResizeJob(wl *kueue.Workload) error
}
```
Jobs implementing the ResizeJob method will create a Workload Slice for every new replica of a job. 

On scale down to M we will change the original Workload's resources and then on scale up to N we will create N-M WorkloadSlice objects that go through admission and scheduling gate checks, i.e. the Workload can "lose" the original (first time admission) quota when scaling down before scaling up.

### Pod Scheduling Gates

Inside **raycluster_webhook** implement schedulingGate injection for pods on RayCluster creation time.
The Pods will be ungated following a similar behavior as to how a job is suspended and then unsuspended in the when admitted.

> When the RayCluster scales up, the new pods will be gated due to the schedulingGates injection in the webhook.

Reviewer 6: Could you make it more generalized, not focusing on the RayCluster?

Author: Sg, i have changed the wording in the Design Details sections to reflect this

> After the creation of each individual Workload Slice and admission of a Workload Slice, the **workload_scheduling_gates_controller** should be in charge of removing the scheduling gates from each pod. All worker pods from the same worker group share the same pod template, so we only need to ungate the number of pods to match the number of admitted pods, this should be a counter. We don’t want to accidentally ungate too many pods since race conditions could happen and we also don’t want to double count. It's worth mentioning that for the case of recreated pods (i.e. machine failure for example), these pods will go through the admission/scheduuling check again, Kueue is responsible fo removing the scheduling gates when there's available quota and resources to spend on the RayCluster. 

Reviewer 6: Do we have only one workload_scheduling_gates_controller? Or do we have such a controller for every framework?

Reviewer 4: I think one should be enough.... We just need to look at the Workload objects and the Pods.

Reviewer 6: > I think one should be enough.... We just need to look at the Workload objects and the Pods.

That makes sense.

### Garbage Collecting Workload Slices

The logic of folding and deleting would be isolated in this controller. We don’t necessarily have to say that this Workload Slice belongs to a specific pod. This controller will look at the Workload objects and check whether they have the Admitted condition or not.

1. The controller increments the `.status.admission.podSetAssignments.count` and passes the UID of the workload you are folding to the parent workload.
If we still don’t see the workload being deleted, we at least know it has been counted towards the parent workload and the controller shouldn't fold it again.

## Phases for MVP (alpha)

### Phase 1 - Scale Down

> Scaling down will be the first phase towards MVP because it can be implemented without introducing the Workload Slices. 

Reviewer 6: Mentioning the differences between the Dynamically Scaling down (this feature) and Dynamically Resource Reclaiming.
As my understanding, the Dynamically Scaling down doesn't release the deleted pod's resource quota, right?

I understand that we currently don't support Dynamically Resource Reclaiming in the RayCluster, but I think that clarifying the differences would be valuable in this phase.

Reviewer 4: dynamic reclaim, for context https://github.com/kubernetes-sigs/kueue/tree/main/keps/78-dynamically-reclaiming-resources

In theory both can work together.

> As my understanding, the Dynamically Scaling down doesn't release the deleted pod's resource quota, right?

It does.

Author: > As my understanding, the Dynamically Scaling down doesn't release the deleted pod's resource quota, right?

@tenzen-y Here's a discussion about that https://github.com/kubernetes-sigs/kueue/pull/1851#discussion_r1537672669

Reviewer 6: > > As my understanding, the Dynamically Scaling down doesn't release the deleted pod's resource quota, right?
> 
> 
> @tenzen-y Here's a discussion about that https://github.com/kubernetes-sigs/kueue/pull/1851#discussion_r1537672669

It seems that the pointing discussion is slightly different.
I meant that can we enable both this feature (KEP-77) and dynamically reclaiming resources (KEP-78).

Reviewer 4: Yes, you could enable both.

- Adding default scheduling gates to control the scheduling of new pods based on their admission status.
- Dynamically adjust quotas in ClusterQueue based on scaling events.

For the MVP, we will only focus on admission of **RayClusters** with autoscaling enabled.

### User Stories

#### Story 1 - RayCluster w/ autoscaling

1. The user creates a RayCluster with `enableInTreeAutoscaling: true`.
2. Kueue admits the RayCluster based on the requested resources in the head pod and worker pod. 
3. Given the demand of the job, the Ray Autoscaler adjusts the replicas field as it adds or removes Pods from the cluster.
4. Kueue will not suspend and requeue the RayCluster, instead it will dynamically update ClusterQueue usage based on the scaling event. 

### Notes/Constraints/Caveats (Optional)

If Kueue needs to preempt the resized RayCluster, it would preempt it as a whole, regardless of whether the RayCluster was previously scaled up.

## Design Details

### Workload Slices

> To support horizontal scaling of jobs, we will introduce the concept of a "Workload Slice”. A Workload Slice is a Workload object  with an owner reference to the original Workload for a job. Workload Slices represent per-replica changes to a job that were not initially accounted for when the job was created. **It's an internal state, not an extra CRD.**

Reviewer 6: ```suggestion
To support horizontal scaling of jobs, we will introduce the concept of a "Workload Slice”. A Workload Slice is a Workload object with an owner reference to the original Workload for a job. Workload Slices represent per-replica changes to a job that were not initially accounted for when the job was created. **It's an internal state, not an extra CRD.**
```

The benefit of Workload Slices is that Kueue can evaluate admission on a per-replica basis without changing the existing semantics of the Workload API. Once a Workload Slice is admitted, it will be garbage collected and its resources will be aggregated into the admission status of the parent workload.
- Workload Slices will be submitted to the same LocalQueue that's referenced by the top-level Workload. 

> - In MultiKueue, Workload Slices would go into the same cluster in a multi-cluster environment.

Reviewer 6: Where is the "same cluster"? Does this mean only a management cluster? Or both management and workload clusters?

Author: Slices will need to go to 2 clusters, management and the worker, i'll update it to reflect this.

  - [Feature Gate](#feature-gate)
  - [Locking Flavor Assignments for Workload Slices](#locking-flavor-assignments-for-workload-slices)
  - [Webhook changes](#webhook-changes)
  - [Test Plan](#test-plan)
    - [Unit Tests](#unit-tests)
    - [Integration tests](#integration-tests)
  - [Graduation Criteria](#graduation-criteria)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
  - [Ignore Resize from Kuberay](#ignore-resize-from-kuberay)
<!-- /toc -->

## Summary

Enable dynamic sizing of Kueue jobs. For the MVP, we will only focus on supporting RayClusters with autoscaling enabled, but other resources that can benefit from dynamic sizing should be supported eventually.

See: [Support dynamically sized (elastic) jobs #77](https://github.com/kubernetes-sigs/kueue/issues/77)

## Motivation

Kueue currently lacks support for resizing jobs. When a job is resized, Kueue will recreate the Workload representation of the job, leading to a disruptive suspend and requeue process. This limitation hinders the usability of Kueue for resources like RayCluster that sometimes have autoscaling capabilities enabled.

> To properly support RayCluster, Kueue needs to gracefully handle the scale up and scale down of RayCluster nodes. Concretely this means that scaling the resources used by a RayCluster is appropriately reflected in the respective ClusterQueue without needing to suspend the entire cluster.

Reviewer 4: ```suggestion
To properly support RayCluster, Kueue needs to gracefully handle the scale up and scale down of RayCluster nodes. Concretely this means that scaling the resources used by a RayCluster is appropriately reflected in the respective ClusterQueue without needing to suspend the entire RayCluster.
```

### Goals

- Gracefully handle resize operations for Kueue jobs (i.e. update quota usage without suspend, enqueue scale ups)

> - Autoscaling RayCluster works with Kueue (MVP)

Reviewer 7: imo an even lower bar MVP could be support with ElasticJobs.

Not supporting JobSet or others is fine but I think we should aim to have both RayCluster and Job..

But I'm not a reviewer so...

Reviewer 2: > ElasticJobs

What is "ElasticJobs"? Is that a custom resource or are you just referring to a Kubernetes Job that can scale horizontally?

Reviewer 7: https://github.com/kubernetes/enhancements/issues/3715

Basically completetions in Jobs are mutable so you should be able to downscale/upscale on a k8s job.

Reviewer 2: Thanks makes sense. I propose we include this in the MVP or a fast follow-up after alpha, but keep RayCluster as a priority since the scope for this smaller / simpler

Reviewer 4: I think RayCluster makes the most sense to start with for 2 reasons:
- it has an autoscaler
- there is high demand

OTOH, HPA doesn't support Job, which in turn reduces demand for elastic Jobs.

Reviewer 7: SGTM.

Reviewer 7: Maybe a comment on the KEP about this could be sufficient. But if its a fast followup its not a big deal.

### Non-Goals

- Vertical scaling of workloads – Kueue will only handle resize operations that scale Pods horizontally 

> - Support resize for other Kueue jobs such as QueuedWorkload, JobSet, etc (future)

Reviewer 4: ```suggestion
- Support resize for other Kueue jobs such as Job, JobSet, etc (future)
```

> - Resizing of RayJobs

Reviewer: What is the practical difference between RayJob and RayCluster in this context?

Author: Ephemeral RayCluster vs long living RayCluster, i.e. defining the size of a cluster needed for a RayJob.

Reviewer 4: From an implementation PoV, it wouldn't be that much harder to support RayJob, given that it's backed by a RayCluster, but there will probably be some nuances.

Reviewer: That is my point. While we probably don't need RayJob in the first implementation PR, I would rather not exclude it completely from the KEP.

Reviewer 4: We can update the KEP once it becomes a goal.

- Partial Preemption

## Proposal

> Update the Job framework reconciler and introduce new controllers to orchestrate dynamic resizing of jobs. We are only interested in horizontal scaling of jobs (e.g. scaling more replicas). At a high level, this will be accomplished by:

Reviewer: Why new controllers are needed (instead of just updating the existing ones)?

Author: To have a clear separation of tasks that are specific to resizing a job: create workload slice(s) -> aggregate workload slice(s) -> ungate pods

- Creating Workload Slice objects that represent incremental scale up of jobs. This gives us per-replica control of workload admission. Workload Slices will be garbage collected and consolidated with their parent workloads after successful admission.
- Adding default scheduling gates to control the scheduling of new pods based on their admission status.
- Dynamically adjust quotas in ClusterQueue based on scaling events.

For the MVP, we will only focus on admission of **RayClusters** with autoscaling enabled.

### User Stories

#### Story 1 - RayCluster w/ autoscaling

1. The user creates a RayCluster.
2. Kueue admits the RayCluster based on the requested resources in the head pod and worker pod. 

> 3. User updates RayCluster to enable autoscaling

Reviewer 4: Couldn't the user directly create the RayCluster with autoscaling enabled too?

Reviewer 2: I think Aldo is talking about just manually updating the replicas of a RayCluster without autoscaling. But I think it's more or less the same user story

Author: Ohh i thought he meant to update the RayCluster CRD and include `enableInTreeAutoscaling: true` as step 3. Nevertheless, I updated the story, PTAL

4. Kueue will not suspend and requeue the RayCluster, instead it will dynamically update ClusterQueue usage based on the scaling event. 

> ## Design Details

Reviewer 4: I don't see a Nodes/caveats section.

One I can think of is that if Kueue needs to preempt the resized RayCluster, it would preempt it as a whole, regardless of whether the RayCluster was previously scaled up.

Author: Okok I added that section.

### Workload Slices
To support horizontal scaling of jobs, we will introduce the concept of a "Workload Slice”. A Workload Slice is a Workload object with an owner reference to the original Workload for a job. Workload Slices represent per-replica changes to a job that were not initially accounted for when the job was created.

The benefit of Workload Slices is that Kueue can evaluate admission on a per-replica basis without changing the existing semantics of the Workload API. Once a Workload Slice is admitted, it will be garbage collected and its resources will be aggregated into the admission status of the parent workload.

### Creating Workload Slices

The [GenericJob interface](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobframework/interface.go#L30-L55) will be updated to handle resize operations of jobs.

```golang
type GenericJob interface {
  ...
  ...
  ResizeJob(wl *kueue.Workload) error
}
```

> Jobs implementing the ResizeJob method will create a Workload Slice for every new replica of a job. 

Reviewer 4: If the implementation of `ResizeJob` is optional, then it should be part of a new interface.

Author: Something like https://github.com/kubernetes-sigs/kueue/blob/b6f7a1c9cfd58fbc25c0909d7910fddb1eba7e8f/pkg/controller/jobframework/interface.go#L60 ?

> Jobs implementing the ResizeJob method will create a Workload Slice for every new replica of a job. 

Author: For now, I don't see how we can include `ResizeJob` as part of the scale down process. Although, it depends on the feedback we receive on #1852 . In fact workloads get constructed, prepared and created inside the job generic reconciler (reconciler.go). So if this method is meant to trigger the creation of a workload slice, i'm not entirely sure if we should delegate this process to happen inside the raycluster_controller for example.

> Jobs implementing the ResizeJob method will create a Workload Slice for every new replica of a job. 

Reviewer 4: What should the implementations of this method do, in general?

> Jobs implementing the ResizeJob method will create a Workload Slice for every new replica of a job. 

Author: My thinking was something similar as [ReclaimablePods](https://github.com/kubernetes-sigs/kueue/blob/c8078b507d3342c0a6c2772f0451ede7a2c04864/pkg/controller/jobframework/reconciler.go#L33) does but in our case we would want to get the replica count difference and the flavor that the top-level workload is using so that we can starting creating the workload slices based on this. Basically, a first step towards the workload slice creation right after the pods are gated.

### Pod Scheduling Gates

> Inside **raycluster_webhook** implement schedulingGate injection for pods on RayCluster creation time. Which will then be ungated following a similar behavior as to how a job is suspended and then unsuspended in the beginning. When we have a scale up, the new pods will be gated due to the schedulingGates injection in the webhook.

Reviewer 4: ```suggestion
Inside **raycluster_webhook** implement schedulingGate injection for pods on RayCluster creation time.
The Pods will be ungated following a similar behavior as to how a job is suspended and then unsuspended in the when admitted.
When the RayCluster scales up, the new pods will be gated due to the schedulingGates injection in the webhook.
```

> After the creation of each individual Workload Slice and admission of a Workload Slice, the **workload_scheduling_gates_controller** should be in charge of removing the scheduling gates from each pod. We only need to ungate the number of pods to match the number of admitted pods, this should be a counter. We don’t want to accidentally ungate too many pods since race conditions could happen and we also don’t want to double count.

Reviewer: What about pods that are just recreated (for example because of machine failure)?

Reviewer 4: Worth mentioning in the doc that the same logic applies for recreated pods.

Author: Added details in this section

### Garbage Collecting Workload Slices

> The logic of folding and deleting would be isolated in this controller that takes a look at the number of pods that were ungated. We don’t necessarily have to say that this Workload Slice belongs to a specific pod. 

Reviewer 4: IIUC, this controller doesn't "take a look at the number of pods that were ungated".

This controller just looks at Workload objects, and whether they are admitted.

> 1. You increment the ungated counter and pass the UID of the workload you are folding to the parent workload. If we still don’t see the workload being deleted, we at least know it has been counted towards the parent workload and we cannot count it again.  

Reviewer 4: ```suggestion
1. The controller increments the `.status.admission.podSetAssignments.count` and passes the UID of the workload you are folding to the parent workload.
If we still don’t see the workload being deleted, we at least know it has been counted towards the parent workload and the controller shouldn't fold it again.
```

> 2. This UID can be seen in the parent’s workload spec. 

Reviewer 4: where exactly? I don't think the spec is a good candidate actually, because you can't update the spec and status at the same time.

Author: Ohhh i meant to say in the parent's `workload.Metadata.UID`

## Phases for MVP (alpha)

> ### Phase 1 - Scale Down

Author: WIP implementation for Phase 1: #1852

Scaling down will be the first phase towards MVP because it can be implemented without introducing the Workload Slices. 

Scaling down a RayCluster won’t involve the creation of Workload Slices, instead it’ll involve an update to the current workload, no requeuing.

1. Compare job's PodSet.Count vs Workload.Spec.PodSets[1].Count (worker group) inside the jobframework generic reconciler. 
2. Call *updateWorkloadToMatchJob()*, this will construct and update the workload and in turn update the PodSet Count field.

> 3. Inside the *Update()* method from the *workload_controller* update the workload in cache and in queue. By updating the workload in cache this will update the cluster queue resource usage and by updating the workload in queue this will trigger the scheduler so that it re-assigns the flavors to the already assumed workload and in this way PodSetAssignments will be updated by applying admission based on the new assignments. 

Reviewer 4: I'm not understanding the notes about the assumed workloads.

An assumed workload is pretty much the same as an admitted workload (kueue is just waiting for the updated event). There wouldn't be a rescheduling attempt for it.

Author: Yes, I guess I was being a bit redundant with the phrasing in this section. This is related to #1852 when an entry for a resize is created, we need to set the `e.status = assumed` and check for this status inside step 5 of the scheduler code (https://github.com/kubernetes-sigs/kueue/blob/536f1ecbbc10d1bd012bd4d8d7c4fb49a7249d28/pkg/scheduler/scheduler.go#L261). Otherwise, it will return an error inside the admit() method, because the workload is already assumed.  We can move the discussion there.

Reviewer 4: ok, that sounds like an implementation detail that shouldn't be part of the KEP.

But also I still don't understand it. I'll review and comment on the implementation PR.

Author: Okok will rephrase this section, thanks

> 4. Inside the schedule logic in the scheduler, since the workload is already assumed in the cache we need to specify if the feature is enabled so that we can apply admission to the workload and update its PodSetAssignments. 

Reviewer 4: This seems unnecessary.

Author: Removed.

#### Job controller

Rework *equivalentToWorkload()* and *reconcile()* to account for potential differences between the job’s number of replicas and the running workload’s PodSet.Count. 

> Given these changes, check if the delta is positive or negative indicating a scaleup/scaledown. If it’s a scaledown the reconciler should trigger an update on the workload to match the new job’s spec and no quota needs to be checked or accounted for, the cluster queue should update the workload resource usage. 

Reviewer 4: And if it's a scale up, we will still suspend and requeue?

Author: Correct

Reviewer 4: ok, please clarify in the doc and explain that it's temporary, while phase 2 and 3 are not completed.

Author: done

### Phase 2 - Aggregating Workload Slices

> In Phase 2, aggregating Workload Slices into the parent workload will be implemented.

Reviewer 4: Add:

This doesn't represent a usable feature to end users, but it can be reviewed independently from the phase 3.

Author: Done.

### Phase 3 - Scale up with Workload Slices and Scheduling Gates

In Phase 3, scale up will be implemented by introducing Workload Slices and adding Pod scheduling gates as part of Kueue’s mutating admission for RayCluster. 

When the RayCluster scales, the RayCluster webhook would be modified to intercept and "gate" all pods. Every time there’s a resize, you create a dependable (child) workload slice and once it's admitted, you increase the count in the original workload, delete the old workload and remove the schedulingGates. 
- Pros: You are able to hold the pods added by Kuberay
- Cons: The fact of having schedulingGates, means we need an API call for every pod, because all pods that are created by the RayCluster are going to have schedulingGates. We need to remove those gates and for every pod you need to make API calls.

#### Scheduler

> Since every scale up will have its own individual workload they should be proposed to the current scheduling cycle and continue the normal admission process. We should lock flavor assignments to Workload Slices we need to ensure that the Workload Slice is assigned the same resource flavor as the parent workload. 

Reviewer 4: ```suggestion
Since every scale up will have its own individual workload they should proceed to the current scheduling cycle and continue the normal admission process.
However, we need to ensure that the Workload Slice is assigned the same resource flavor as the parent workload. 
```

> The *nominate()* returns the workloads with their requirements (resource flavors, borrowing) if they were admitted by the clusterQueues in the snapshot, so we need to return the original workload that was already admitted with the resize information inside *TotalRequests* so that *PodSetAssignments* is also updated.

Reviewer 4: ```suggestion
The `nominate()` returns the workloads with their requirements (resource flavors, borrowing) if they were admitted by the clusterQueues in the snapshot, so we need to return the original workload that was already admitted with the resize information inside *TotalRequests* so that *PodSetAssignments* is locked to the parent Workload assignments.
```

## Additional Details

### Feature Gate

> In kube_features add Elastic/Dynamic size jobs feature gate.

Reviewer 4: No need to mention this. It's already mentioned in kep.yaml

Author: Sg, removed

> ### Locking Flavor Assignments for Workload Slices

Reviewer 4: Unnecessary implementation detail.

Author: Removed

We can extract the flavor(s) that the parent workload is using through wl.Status.Admission.PodSetAssigments[1].Flavors

We add a new field to the Workload Info object to know which parent flavor(s) were used.

```golang
// Info holds a Workload object and some pre-processing.
type Info struct {
   Obj *kueue.Workload
   // list of total resources requested by the podsets.
   TotalRequests []PodSetResources
   // Populated from the queue during admission or from the admission field if
   // already admitted.
   ClusterQueue   string
   LastAssignment *AssignmentClusterQueueState
 // Parent Flavors
 ParentFlavor []string
}
```

Which can be passed on as an extra parameter in *getAssignments()* in the scheduler to flavorassigner.go when it assigns a flavor through *assignFlavors()*

```golang
func (a *FlavorAssigner) assignFlavors(log logr.Logger, requests []workload.PodSetResources, ParentFlavor []string) Assignment {}
```

Which then calls *findFlavorForPodSetResource()* and we can use the parent flavor(s) value to check if this flavor can fit to the pod. If it doesn’t we don’t try to find another flavor for it.

> ### Webhook changes

Reviewer 4: Also unnecessary implementation detail

Author: Removed

        allErrs = append(allErrs, apivalidation.ValidateImmutableField(newObj.Spec.PriorityClassSource, oldObj.Spec.PriorityClassSource, specPath.Child("priorityClassSource"))...)
        allErrs = append(allErrs, apivalidation.ValidateImmutableField(newObj.Spec.PriorityClassName, oldObj.Spec.PriorityClassName, specPath.Child("priorityClassName"))...)
    }
...
- allErrs = append(allErrs, validateAdmissionUpdate(newObj.Status.Admission, oldObj.Status.Admission, field.NewPath("status", "admission"))...)

```

### Test Plan

<!--
**Note:** *Not required until targeted at a release.*
The goal is to ensure that we don't accept enhancements with inadequate testing.

All code is expected to have adequate tests (eventually with coverage
expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
when drafting this test plan.

[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
-->

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

> #### Unit Tests

Reviewer 4: Please fill this up.

Author: Added more details

to implement this enhancement and provide the current unit coverage for those
in the form of:
- <package>: <date> - <current test coverage>

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
-->

#### Integration tests

### Graduation Criteria
<!--

Clearly define what it means for the feature to be implemented and
considered stable.

If the feature you are introducing has high complexity, consider adding graduation
milestones with these graduation criteria:
- [Maturity levels (`alpha`, `beta`, `stable`)][maturity-levels]
- [Feature gate][feature gate] lifecycle
- [Deprecation policy][deprecation-policy]

[feature gate]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
-->

The feature starts at the alpha level, with a feature gate.

In the Alpha version, Dynamically Sized Jobs will support RayCluster resizing. 

> - Scale down/up of replica workers

Reviewer 4: Give a bit more details on the kind of scenarios that you will test.

Something worth testing is preemption: what if the parent Workload is preempted when there are unmerged Workload slices?

Author: I added more general scenarios to test in the Integration Tests section
